File: 剑指OfferWithJava/src/Sort/MergeSort.java
package Sort;

import java.util.Arrays;

@SuppressWarnings({"all"})
public class MergeSort {
    public static void mergeSort(int[] nums) {
        if (nums == null || nums.length <2) return;
        mergeSort(nums, 0, nums.length-1);
    }
    public  static void mergeSort(int[] nums, int l, int r) {
        if(l>=r) return;
//        while (l<r) {

        int mid = l + ((r - l) >> 1);
        mergeSort(nums, l, mid);
        mergeSort(nums, mid + 1, r);
        merge(nums,l,mid,r);
//        }
    }
    public static void merge(int[] nums, int l, int mid,int r) {
        int p1=l,p2=mid+1,i=0;
        int[] helper = new int[r - l + 1];
        while (p1<=mid && p2 <= r ) {
            helper[i++] = nums[p1]<nums[p2] ?nums[p1++] : nums[p2++];
        }
        while (p1 <=mid) {
            helper[i++] = nums[p1++];
        }
        while (p2<=r) {
            helper[i++] = nums[p2++];
        }
        for ( i = 0; i < helper.length; i++) {
            nums[i+l] = helper[i];
        }
    }

    public static void main(String[] args) {
        int[] nums = {1, 0, 9, 8, 7, 2, 3, 6, 5, 4, 13, 11, 12};
        mergeSort(nums);
        System.out.println("Arrays.toString(nums) = " + Arrays.toString(nums));
    }
}

File: LC-Go/Sorts/MergeSort.go
package Sorts

func MergeSort(nums []int) []int {
	if nums == nil {
		return nums
	}
	mergeSort(nums, 0, len(nums)-1)
	return nums
}

func mergeSort(nums []int, l int, r int) {
	if l >= r {
		return
	}
	mid := l + ((r - l) >> 1)
	mergeSort(nums, l, mid)
	mergeSort(nums, mid+1, r)
	// 传递引用 此处不是内部递归调用 调用 merge
	merge(&nums, l, mid, r)

}

func merge(nums *[]int, l int, mid int, r int) {

	p1, p2 := l, mid+1
	helper := make([]int, r-l+1)
	i := 0
	for p1 <= mid && p2 <= r {
		if (*nums)[p1] <= (*nums)[p2] {
			helper = append(helper, (*nums)[p1])
			p1++
		} else {
			helper = append(helper, (*nums)[p2])
			p2++
		}
		i++
	}
	for j := 0; j < i; j++ {
		(*nums)[j+l] = helper[j]
	}
}
